### PR TITLE
Add gravatar support

### PR DESCRIPTION
**Pull request by daquexian**
Try to load gravatar as contact's picture as fallback. And a option in settings is added.

However it is hard to get `Account` object in `ContactPictureLoader`, so signatures of many methods are modified for passing the `Account` object to `ContactPictureLoader`. It will be better if there is a way to get `Account` object directly anywhere.

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java [hunk 1]
```diff
@@ -271,4 +332,24 @@ public void cancel() {
         }
     }
 
+    private Uri getGravatarUri(String email) {
+        return Uri.parse("http://www.gravatar.com/avatar/" + getMD5String(email) + "?d=404");
+    }
+
+    private String getMD5String(String from) {
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java]: we have [okio](http://square.github.io/okio/1.x/okio/) in our dependencies, there should be `ByteBuffer.md5()` available for this purpose

**Comment by cketti** [comment]
Still not a fan of including this directly. See the comments in #502.

File: k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java [hunk 2]
```diff
@@ -138,16 +149,53 @@ private void loadFallbackPicture(Address address, ImageView imageView) {
...
+                    @Override
+                    public boolean onResourceReady(GlideDrawable resource, FallbackGlideParams model,
+                            Target<GlideDrawable> target,
+                            boolean isFromMemoryCache, boolean isFirstResource) {
+                        return false;
+                    }
+                })
                 .into(imageView);
+
+        if (account.getUseGravatar() == UseGravatar.ALWAYS || (account.getUseGravatar() == UseGravatar.ON_WIFI && isOnWiFi(imageView.getContext()))) {
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java]: Line is too long

File: k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java [hunk 3]
```diff
@@ -271,4 +332,24 @@ public void cancel() {
         }
     }
 
+    private Uri getGravatarUri(String email) {
+        return Uri.parse("http://www.gravatar.com/avatar/" + getMD5String(email) + "?d=404");
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/misc/ContactPictureLoader.java]: `"http://www.gravatar.com/avatar/%s?d=404"` should be a constant format string defined somewhere reasonable (possibly `K9.GRAVATAR_URI_FORMAT` or in the XML).

**Comment by Valodim** [comment]
I disagree about the "separate app" idea. The complexity of even just a gravatar app would be an order of magnitude higher than this 150 lines implementation. On top of that, users really don't dig the "pluggable apps" idea at all, even with OpenKeychain I'm sometimes having doubts whether it's the way to go.

If people really want the configurability, a setting that takes a stringformat with some placeholders and preconfigured values would work just fine. Personally, I don't even know of any other services someone might want to use, and would assume gravatar covers the 99% case.

All in all, I think this strikes a reasonable effort/usefulness tradeoff.

**Comment by cketti** [comment]
I agree that the complexity of a separate app is magnitudes higher. Which is probably the reason why nobody bothered to implement that idea. That's totally fine with me.
Gravatar is not a core email feature. It's not even a reasonably important auxiliary feature. I'm all for polishing the app with things like that once we got all the important bits right. Before that the code is just a burden. Even the stupid dynamic fallback image has caused more trouble than it should have.

Also, I'm not excited about the implementation. `ContactPictureLoader` wasn't great before. Now it's coupled to `Account` and `ConnectivityManager`. And there's also the "easy" stuff: `e.printStackTrace()`, `System.out.println()` :(

I strongly believe that not including this feature now will help K-9 Mail in the long run.

**Comment by daquexian** [comment]
Frankly speaking I'm not excited about the implementation as well. Signatures of many functions have to be modified to retrieve the setting. If there is a way to obtain `Account` object directly the code will be much neater.

About the necessity of Gravatar, it's obvious that I have much less knowledge about K-9 than you maintainers, so I will obey your decision if there is a unified opinion after discussion :P

**Pull request closed without merging**